bytecomp/matching.ml

```diff
@@ -1178,160 +1193,66 @@ let rec split_or argo cls args def =
-          else
-            let yes = List.rev rev_yes in
-            let { me = next; matrix; top_default = def }, nexts =
-              split_exc cstr [ cl ] rem
-            in
-            let idef = next_raise_count () in
-            let def = Default_environment.cons matrix idef def in
-            ( { me = Pm { cases = yes; args; default = def };
-                matrix = as_matrix yes;
-                top_default = def
-              },
-              (idef, next) :: nexts )
-        else
-          let yes = List.rev rev_yes in
-          let { me = next; matrix; top_default = def }, nexts =
-            split_noexc [ cl ] rem
-          in
-          let idef = next_raise_count () in
-          let def = Default_environment.cons matrix idef def in
-          ( { me = Pm { cases = yes; args; default = def };
-              matrix = as_matrix yes;
-              top_default = def
-            },
-            (idef, next) :: nexts )
-    | _ -> assert false
-  and split_noexc rev_yes = function
+and split_no_or cls args def k =
+  let rec split cls =
+    let discr = what_is_first_case cls in
+    collect discr [] [] cls
+  and collect group_discr rev_yes rev_no = function
     | ([], _) :: _ -> assert false
-    | [] -> precompile_var args (List.rev rev_yes) def k
-    | ((p :: _, _) as cl) :: rem ->
-        if not (group_constructor p) then
-          split_noexc (cl :: rev_yes) rem
+    | [ ((ps, _) as cl) ] when List.for_all group_var ps && rev_yes <> [] ->
+        (* This enables an extra division in some frequent cases:
+               last row is made of variables only *)
+        collect group_discr rev_yes (cl :: rev_no) []
+    | ((p :: _, _) as cl) :: rem -> (
+        if can_group group_discr p && safe_before cl rev_no then
+          collect group_discr (cl :: rev_yes) rev_no rem
         else
-          let yes = List.rev rev_yes in
-          let { me = next; matrix; top_default = def }, nexts =
-            split_exc (pat_as_constr p) [ cl ] rem
-          in
-          let idef = next_raise_count () in
-          precompile_var args yes
-            (Default_environment.cons matrix idef def)
-            ((idef, next) :: nexts)
-  in
-  match cls with
-  | [] -> assert false
-  | ((p :: _, _) as cl) :: rem ->
-      if group_constructor p then
-        split_exc (pat_as_constr p) [ cl ] rem
+          (* If we can't group the current clause, we have on more decision to
+             make: do we want to try raising lower clauses, or do we want to
+             split now?
+             The decision is made based on the following heuristic: if the group
+             consists of extension constructors, it is unlikely that we are
+             going to raise anything, so we split now.
+             Otherwise, we try to raise lower clauses. *)
+          match group_discr.pat_desc with
+          | Tpat_construct (_, { cstr_tag = Cstr_extension _ }, _) ->
+              assert (rev_no = []);
```

Comment:
I'm still not fond of the way split-global coherence assumptions are interleaved with clause-local decision-making here, resulting in this assertion.